### ISSUE TITLE
Question: "No JWT found" for tomcat service in workspace

### ISSUE THREAD

**Issue by sonntag-dev** [eclipse-che/che#16503]
Hi,

I have a multi user Eclipse Che environment. In this environment only the user of a specific workspace should have access to the tomcat service within the workspace.
For example, as long as the service in my workspace is accessible by url everybody that knows this url has access to this service, too. (If she or he has access to the server/cluster where my eclipse che is running)
Therefore I would like to use the JWT proxy for my tomcat services for each separate workspace in Eclipse Che. 
So I added _"secure":"true"_ for my service endpoint in my devfile to enable JWT proxy. But when I want to open the service with the editor within my workspace 

![image](https://user-images.githubusercontent.com/60471467/78047976-560ab200-7379-11ea-8645-7eccc2251cad.png)

I get the following message: **jwtproxy: unable to verify request: No JWT found**

In #10252 there is an explanation that I can use the "CHE_MACHINE_TOKEN" of my runtime container as token. I tested it with postman, added the CHE_MACHINE_TOKEN in the header and it works.  

But is it possible to open the service directly with the hyperlink in the right sidebar within the editor (see image), if JWT proxy is enabled? I think it is really handy to jump directly to the service with the help of this link.
Do I have to change my devfile or how can I pass the CHE_MACHINE_TOKEN within this link to authorize against JWT proxy?

Thank you for your help.

**Comment by skabashnyuk** [comment]
@sonntag-dev thanks for reporting this issue.  Can you share devfile to test this?

**Comment by sleshchenko** [comment]
@sonntag-dev After you added `secure: true` - the token must be sent for each request, so it's expected. Che contains an automated mechanism for such token to be sent, it's `cookiesAuthEnabled` server's attribute. See https://github.com/eclipse/che-plugin-registry/blob/master/v3/plugins/eclipse/che-theia/7.11.0/meta.yaml#L22
But note that after that you might think about CSRF(Cross-Site Request Forgery), we don't care about it with Theia because we made sure that Theia does not have modifying APIs methods, only non-modifying or WebSocket.
^ It's the reason why `cookiesAuthEnabled: false` by default, server developer must make sure that their server has additional CSRF protection or this issue is not actual for their server.

**Comment by sonntag-dev** [comment]
I used Che version 7.8.
After I added the cookiesAuthEnabled server attribute my devfile looks like this:

```
metadata:
  generateName: Test-
projects:
  - name: vertx-http-example
    source:
      location: 'https://github.com/openshift-vertx-examples/vertx-http-example.git'
      type: git
      branch: master
attributes:
  persistVolumes: 'true'
components:
  - mountSources: true
    endpoints:
      - name: client
        port: 8080
        attributes:
          cookiesAuthEnabled: 'true'
          path: <path>
          secure: 'true'
    memoryLimit: 4Gi
    type: dockerimage
    volumes:
      - name: volume1
        containerPath: <containerPathVolume1>
      - name: volume2
        containerPath: <containerPathVolume2>
      - name: volume3
        containerPath: <containerPathVolume3>
    image: '<privateDockerImageRepository>'
    alias: application
  - id: redhat/java/latest
    type: chePlugin
  - mountSources: true
    endpoints:
      - name: postgres-sql-database
        port: 5432
        attributes:
          secure: 'true'
    memoryLimit: 512Mi
    type: dockerimage
    volumes:
      - name: postgres-persistent-storage
        containerPath: <postgres-persistent-storage-containerPath>
    image: 'postgres:11'
    alias: postgres-sql-database
    env:
      - value: <password>
        name: POSTGRES_PASSWORD
      - value: <user>
        name: POSTGRES_USER
apiVersion: 1.0.0
commands:
  - name: Debug remote java application
    actions:
      - referenceContent: |
          {
          "version": "0.2.0",
          "configurations": [
            {
              "type": "java",
              "name": "Debug (Attach) - Remote",
              "request": "attach",
              "hostName": "localhost",
              "port": 5005
            }]
          }
        type: vscode-launch
```

I changed my Che configMap to:

```
apiVersion: v1
data:
  CHE_API: http://<fqdn>/api
  CHE_CORS_ALLOW__CREDENTIALS: "true"
  CHE_CORS_ALLOWED__ORIGINS: "http://<fqdn>"
  CHE_CORS_ENABLED: "true"
  CHE_DEBUG_SERVER: "true"
  CHE_HOST: "<fqdn>"
  ........
```
FQDN is the domain where my che host service is running.

But I still get the message that the access to XMLHttpRequest  has been blocked by CORS policy, because no 'Access-Control-Allow-Origin' header is present on the requested resource.

Is there a problem with my server side configuration?

Thank you for your help.

**Comment by skabashnyuk** [comment]
>  CHE_CORS_ALLOW__CREDENTIALS: "true"
  CHE_CORS_ALLOWED__ORIGINS: "http://<fqdn>"
  CHE_CORS_ENABLED: "true"

I would not recommend enabling CORS for production use at all.

>But I still get the message that the access to XMLHttpRequest has been blocked by CORS policy, because no 'Access-Control-Allow-Origin' header is present on the requested resource.

Can you specify:
- from what application it comes from.
- The application is doing a request to wich endpoint.
Ideally, can you show Chrome dev console error?



can you use https://tomcat.apache.org/tomcat-8.5-doc/realm-howto.html#UserDatabaseRealm to authorize users to you application inside of the workplace?

**Comment by sleshchenko** [comment]
@sonntag-dev Also could you try to use `/` as path. We have known issue that che-jwtproxy  + wsloader does not work correctly with non-root path https://github.com/eclipse/che/issues/15867

**Comment by sonntag-dev** [comment]
This is the console error:

![image](https://user-images.githubusercontent.com/60471467/78341228-a1021080-7597-11ea-8bf7-e6d602c29ea3.png)

I changed the Che configMap to:

  ```
  CHE_CORS_ALLOW__CREDENTIALS: "false"
  CHE_CORS_ALLOWED__ORIGINS: '*'
  CHE_CORS_ENABLED: "false"
```
and redeploy the Che deployment. But I still got the CORS error.

Then I ran Chrome with disabled CORS and got this console error:

![image](https://user-images.githubusercontent.com/60471467/78342952-677ed480-759a-11ea-9e4c-7a161ed0c580.png)

A look at my ingress controller in k8s cluster shows me that there are 4 different controller for JWT proxy:

- servergl1erhj1-jwtproxy-server-4400.......
- servergl1erhj1-jwtproxy-server-4401.......
- servergl1erhj1-jwtproxy-server-4402.......
- servergl1erhj1-jwtproxy-server-4403.......

**Comment by skabashnyuk** [comment]
Did you see the preflight request in the console? What was the response?

**Comment by sonntag-dev** [comment]
Preflight request:
- origin: http://<fqdn>
- Access-Control-Request-Headers: authorization
- Access-Control-Request-Method: GET
- Host: servergl1erhj1-jwtproxy-server-4401.<domainName>

Response CORS enabled:
- Status: 403

Response CORS disabled:
- Status: 404

I checked 
`http://servergl1erhj1-jwtproxy-server-4401.<domainName>/<endPointPath> `
in Postman with the CHE_MACHINE_TOKEN as authorization bearer token directly and I got a status 200 response. So the endpoint is accessible with CHE_MACHINE_TOKEN and Postman.

I also see that in the loader.js is checked for status 204:

![image](https://user-images.githubusercontent.com/60471467/78377738-49cc6200-75d0-11ea-9f37-4dac73c1a559.png)

With another service I tried when CORS is disabled I got response status 200 but it throws an exception:

![image](https://user-images.githubusercontent.com/60471467/78378124-d119d580-75d0-11ea-9f37-14bac9823bdb.png)

Why does the loader check for response status 204? Does anyone has an idea why the proxy redirection didn't find my service endpoints?

The service with the status 200 response has the same devfile endpoint definition than all other services.
